go/runtime/registry/registry.go

```diff
@@ -494,20 +495,21 @@ func (r *runtimeRegistry) addSupportedRuntime(ctx context.Context, id common.Nam
 		return fmt.Errorf("runtime/registry: runtime already registered: %s", id)
 	}
 
-	path, err := EnsureRuntimeStateDir(r.dataDir, id)
-	if err != nil {
-		return err
-	}
-
-	rt, err := newRuntime(ctx, id, r.cfg, r.consensus, r.logger)
+	rt, err := newRuntime(ctx, r.dataDir, id, r.cfg, r.consensus, r.logger)
 	if err != nil {
 		return err
 	}
+	defer func() {
+		if rerr != nil {
+			rt.localStorage.Stop()
```

Comment:
I've changed it a bit so that it now calls `rt.stop()` in case of failures. This avoids requiring callers to know about internals (but they still need to know that they should call stop in case of errors after `newRuntime`).

Reply:
I think that the right procedure would be to call runtime's `stop`, like you pointed. But the problem is that history is constructed outside the `newRuntime`. Something like this:
```
func newRuntime() {
	...
	r.s, err = storage.New()
	defer r.s.close() // TODO: close only on error 
	r.h, err := history.New()
	defer r.h.close() // TODO: close only on error 
	...
	return r, nil
}

func someFunc() {
	r = newRuntime()
	defer r.close() // TODO: close only on error if function returns runtime
}
```

Reply:
Yeah, but `newRuntime` does not create the history as some runtimes do not need it (while all need local storage). As mentioned I've updated it now so that one should just call `rt.close()` in such cases and it will work whether the history is there or not.